File: AEDS-II/verde/TP02/11/Show.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Show{
    
    private String SHOW_ID;
    private String TYPE;
    private String TITLE;
    private String DIRECTOR;
    private String[] CAST;
    private String COUNTRY;
    private String DATE_ADDED;
    private int RELEASE_YEAR;
    private String RATING;
    private String DURATION;
    private String[] LISTED_IN;

    public Show(){
        
        this.SHOW_ID = this.TYPE = this.TITLE = this.DIRECTOR = this.COUNTRY = this.DATE_ADDED = this.RATING = this.DURATION = "NaN";
        this.CAST = new String[0];
        this.LISTED_IN = new String[0];
        this.RELEASE_YEAR = 0;
    }

    public Show(String SHOW_ID, String TYPE, String TITLE, String DIRECTOR, String[] CAST,
                String COUNTRY, String DATE_ADDED, int RELEASE_YEAR, String RATING,
                String DURATION, String[] LISTED_IN){
        
        this.SHOW_ID = SHOW_ID;
        this.TYPE = TYPE;
        this.TITLE = TITLE;
        this.DIRECTOR = DIRECTOR;
        this.CAST = CAST;
        this.COUNTRY = COUNTRY;
        this.DATE_ADDED = DATE_ADDED;
        this.RELEASE_YEAR = RELEASE_YEAR;
        this.RATING = RATING;
        this.DURATION = DURATION;
        this.LISTED_IN = LISTED_IN;
    }

    public Show clone(){
        return new Show(SHOW_ID, TYPE, TITLE, DIRECTOR, CAST.clone(), COUNTRY, DATE_ADDED, RELEASE_YEAR, RATING, DURATION, LISTED_IN.clone());
    }

    public void imprimir(){
        
        System.out.print("=> " + SHOW_ID + " ## " + TITLE + " ## " + TYPE + " ## " + DIRECTOR + " ## ");
        System.out.print(Arrays.toString(CAST) + " ## " + COUNTRY + " ## " + DATE_ADDED + " ## ");
        System.out.print(RELEASE_YEAR + " ## " + RATING + " ## " + DURATION + " ## ");
        System.out.println(Arrays.toString(LISTED_IN) + " ##");
    }

    public void ler(String linha){

        String[] campos = new String[12];
        int pos = 0;
        boolean dentroAspas = false;
        StringBuilder campo = new StringBuilder();

        for(int i = 0; i < linha.length(); i++){
            char c = linha.charAt(i);
            if(c == '\"') dentroAspas = !dentroAspas;
            
            else if(c == ',' && !dentroAspas){
                campos[pos++] = campo.toString().trim();
                campo.setLength(0);
            }
            else campo.append(c);
        }
        
        campos[pos] = campo.toString().trim();

        this.SHOW_ID = campos[0].isEmpty() ? "NaN" : campos[0];
        this.TYPE = campos[1].isEmpty() ? "NaN" : campos[1];
        this.TITLE = campos[2].isEmpty() ? "NaN" : campos[2];
        this.DIRECTOR = campos[3].isEmpty() ? "NaN" : campos[3];
        this.CAST = campos[4].isEmpty() ? new String[]{ "NaN" } : campos[4].split(", ");
        Arrays.sort(this.CAST);
        this.COUNTRY = campos[5].isEmpty() ? "NaN" : campos[5];
        this.DATE_ADDED = campos[6].isEmpty() ? "March 1, 1900" : campos[6];
        
        try{
            this.RELEASE_YEAR = Integer.parseInt(campos[7]);
        } 
        catch (Exception e){
            this.RELEASE_YEAR = 0;
        }

        this.RATING = campos[8].isEmpty() ? "NaN" : campos[8];
        this.DURATION = campos[9].isEmpty() ? "NaN" : campos[9];
        this.LISTED_IN = campos[10].isEmpty() ? new String[]{ "NaN" } : campos[10].split(", ");
        Arrays.sort(this.LISTED_IN);
    }

    public int getRELEASE_YEAR(){
        return this.RELEASE_YEAR;
    }

    public String getTITLE(){
        return this.TITLE;
    }

    public static void main(String[] args) throws Exception{
        
        BufferedReader teclado = new BufferedReader(new InputStreamReader(System.in));
        String linha;
        String[] entradas = new String[1000];
        int n = 0;

        while(!(linha = teclado.readLine()).equals("FIM")){
            entradas[n++] = linha;
        }

        String[] csv = new String[10000];
        int m = 0;
        
        BufferedReader br = new BufferedReader(new FileReader("/tmp/disneyplus.csv"));
        ///home/vicenzo/Documentos/GitHub/Lista-AEDS/AEDS-II/verde/TP02/disneyplus.csv

        br.readLine();
        
        while((linha = br.readLine()) != null){
            csv[m++] = linha;
        }
        br.close();

        Show[] lista = new Show[n];
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(csv[j].startsWith(entradas[i] + ",")){
                  
                    lista[i] = new Show();
                    lista[i].ler(csv[j]);
                    break;
                }
            }
        }

        //counting
        long inicio = System.nanoTime();
        int comp = 0, mov = 0;

        int minAno = Integer.MAX_VALUE;
        int maxAno = Integer.MIN_VALUE;

        for(int i = 0; i < n; i++){
            if (lista[i].RELEASE_YEAR < minAno) minAno = lista[i].RELEASE_YEAR;
            if (lista[i].RELEASE_YEAR > maxAno) maxAno = lista[i].RELEASE_YEAR;
        }

        int range = maxAno - minAno + 1;
        int[] count = new int[range];
        
        for (int i = 0; i < n; i++){
            count[lista[i].RELEASE_YEAR - minAno]++;
        }

        for(int i = 1; i < range; i++){
            count[i] += count[i - 1];
        }

        Show[] ordenado = new Show[n];
        
        for(int i = n - 1; i >= 0; i--){
            int pos = count[lista[i].RELEASE_YEAR - minAno] - 1;
            ordenado[pos] = lista[i];
            count[lista[i].RELEASE_YEAR - minAno]--;
            mov++;
        }

        //Desempate
        for(int i = 1; i < n; i++){
            Show tmp = ordenado[i];
            int j = i - 1;
            
            while(j >= 0 && ordenado[j].RELEASE_YEAR == tmp.RELEASE_YEAR &&
                   ordenado[j].TITLE.compareTo(tmp.TITLE) > 0){
                
                ordenado[j + 1] = ordenado[j];
                j--;
                comp++;
                mov++;
            }

            ordenado[j + 1] = tmp;
            mov++;
        }

        for(int i = 0; i < n; i++){
            ordenado[i].imprimir();
        }

        long fim = System.nanoTime();

        java.io.FileWriter fw = new java.io.FileWriter("matricula_countingsort.txt");
        fw.write("802305\t" + comp + "\t" + mov + "\t" + (fim - inicio) / 1000000 + "\n");
        fw.close();
    }
}

File: AEDS-II/verde\tps/TP02/16/Show.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

#define MAX 1000
#define K 10

typedef struct {
    char show_id[20];
    char type[100];
    char title[300];
    char director[200];
    char cast[500];
    char country[100];
    char date_added[100];
    int release_year;
    char rating[20];
    char duration[50];
    char listed_in[300];
} Show;

void trim(char *str) {
    char *start = str;
    while (*start == ' ' || *start == '\n' || *start == '\r') start++;
    if (start != str) memmove(str, start, strlen(start) + 1);

    char *end = str + strlen(str) - 1;
    while (end > str && (*end == ' ' || *end == '\n' || *end == '\r')) *end-- = '\0';
}

void tratarCampo(char *campo) {
    trim(campo);
    if (strlen(campo) == 0) strcpy(campo, "NaN");
}

void lerShow(char *linha, Show *s) {
    char *token;
    token = strtok(linha, ",");
    strcpy(s->show_id, token ? token : "NaN");

    token = strtok(NULL, ",");
    strcpy(s->type, token ? token : "NaN");

    token = strtok(NULL, ",");
    strcpy(s->title, token ? token : "NaN");

    token = strtok(NULL, ",");
    strcpy(s->director, token ? token : "NaN");

    token = strtok(NULL, ",");
    strcpy(s->cast, token ? token : "NaN");

    token = strtok(NULL, ",");
    strcpy(s->country, token ? token : "NaN");

    token = strtok(NULL, ",");
    if (token) trim(token);
    strcpy(s->date_added, (token && strlen(token) > 0) ? token : "March 1, 1900");

    token = strtok(NULL, ",");
    s->release_year = token ? atoi(token) : 0;

    token = strtok(NULL, ",");
    strcpy(s->rating, token ? token : "NaN");

    token = strtok(NULL, ",");
    strcpy(s->duration, token ? token : "NaN");

    token = strtok(NULL, ",");
    strcpy(s->listed_in, token ? token : "NaN");

    tratarCampo(s->show_id);
    tratarCampo(s->type);
    tratarCampo(s->title);
    tratarCampo(s->director);
    tratarCampo(s->cast);
    tratarCampo(s->country);
    tratarCampo(s->date_added);
    tratarCampo(s->rating);
    tratarCampo(s->duration);
    tratarCampo(s->listed_in);
}

void imprimirShow(Show s) {
    printf("=> %s ## %s ## %s ## %s ## [%s] ## %s ## %s ## %d ## %s ## %s ## [%s] ##\n",
           s.show_id, s.title, s.type, s.director, s.cast, s.country, s.date_added,
           s.release_year, s.rating, s.duration, s.listed_in);
}

void insertionSort(Show array[], int n, int *comp, int *mov) {
    for (int i = 1; i < n; i++) {
        Show temp = array[i];
        int j = i - 1;
        while (j >= 0 && (array[j].release_year > temp.release_year ||
                         (array[j].release_year == temp.release_year &&
                          strcmp(array[j].title, temp.title) > 0))) {
            (*comp)++;
            array[j + 1] = array[j];
            (*mov)++;
            j--;
        }
        (*comp)++;
        array[j + 1] = temp;
        (*mov)++;
    }
}

int main() {
    FILE *fp = fopen("/tmp/disneyplus.csv", "r");
    if (!fp) {
        printf("Erro ao abrir o arquivo.\n");
        return 1;
    }

    Show lista[K];
    int count = 0;
    char linha[1000];
    char entrada[20];
    char *todasLinhas[MAX];
    int total = 0;

    // Ler todas as linhas do CSV
    while (fgets(linha, sizeof(linha), fp) && total < MAX) {
        todasLinhas[total] = strdup(linha);
        total++;
    }
    fclose(fp);

    // Ler entradas
    while (scanf("%s", entrada) && strcmp(entrada, "FIM") != 0 && count < K) {
        for (int i = 0; i < total; i++) {
            if (strstr(todasLinhas[i], entrada)) {
                lerShow(todasLinhas[i], &lista[count++]);
                break;
            }
        }
    }

    clock_t inicio = clock();
    int comp = 0, mov = 0;
    insertionSort(lista, count, &comp, &mov);
    clock_t fim = clock();

    for (int i = 0; i < count; i++) {
        imprimirShow(lista[i]);
    }

    double tempo = (double)(fim - inicio) / CLOCKS_PER_SEC;
    FILE *log = fopen("matricula_insercao.txt", "w");
    fprintf(log, "812313	%d	%d	%.6f", comp, mov, tempo);
    fclose(log);

    for (int i = 0; i < total; i++) free(todasLinhas[i]);
    return 0;
}
